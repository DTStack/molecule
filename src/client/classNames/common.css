*:focus-visible {
    outline: none;
}

body [tabindex='0']:focus,
body [tabindex='-1']:focus,
body .synthetic-focus,
body select:focus,
body input[type='button']:focus,
body input[type='text']:focus,
body button:focus,
body textarea:focus,
body input[type='search']:focus,
body input[type='checkbox']:focus {
    outline-width: 1px;
    outline-style: solid;
    outline-offset: -1px;
    outline-color: var(--focusBorder);
    opacity: 1;
}

body .synthetic-focus :focus {
    outline: 0 !important; /* elements within widgets that draw synthetic-focus should never show focus */
}

body .monaco-inputbox.info.synthetic-focus,
body .monaco-inputbox.warning.synthetic-focus,
body .monaco-inputbox.error.synthetic-focus,
body .monaco-inputbox.info input[type='text']:focus,
body .monaco-inputbox.warning input[type='text']:focus,
body .monaco-inputbox.error input[type='text']:focus {
    outline: 0 !important; /* outline is not going well with decoration */
}

body .monaco-list:focus {
    outline: 0 !important; /* tree indicates focus not via outline but through the focused item */
}

body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Noto Sans', Helvetica, Arial,
        sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji';
}
