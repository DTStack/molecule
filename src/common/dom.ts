export type HTMLElementType = HTMLElement | null;
export type TriggerEvent = 'click' | 'contextmenu' | 'hover';
/**
 * specify `rightBottom` means align to the bottom and keep in right
 */
export type PlacementType = 'top' | 'right' | 'bottom' | 'left' | 'rightBottom';
export const select = document.querySelector.bind(document);
export const selectAll = document.querySelectorAll.bind(document);

export interface IPosition {
    x: number;
    y: number;
}

/**
 * Get Document Rectangle info
 */
export function getDocumentRect() {
    const body = document.body;
    const html = document.documentElement;

    const clientWidth = html.clientWidth;
    const clientHeight = html.clientHeight;

    const height = Math.max(
        body.scrollHeight,
        body.offsetHeight,
        clientHeight,
        html.scrollHeight,
        html.offsetHeight
    );

    const width = Math.max(
        body.scrollWidth,
        body.offsetWidth,
        clientWidth,
        html.scrollWidth,
        html.offsetWidth
    );

    return {
        height,
        width,
        clientWidth,
        clientHeight,
    };
}

/**
 * Returns the position of element relative to element position
 * @param element target element
 * @param relativePos the relative element position
 */
export function getRelativePosition(
    element: HTMLElement,
    relativePos: IPosition
) {
    const { clientWidth, clientHeight } = getDocumentRect();
    const anchorX = relativePos.x;
    const anchorY = relativePos.y;

    const marginRight = clientWidth - anchorX;
    const marginBottom = clientHeight - anchorY;

    const viewHeight = element.offsetHeight;
    const viewWidth = element.offsetWidth;

    const x = marginRight < viewWidth ? anchorX - viewWidth : anchorX;
    const y = marginBottom < viewHeight ? anchorY - viewHeight : anchorY;

    return {
        x,
        y,
    };
}

export function getEventPosition(e: React.MouseEvent) {
    return {
        x: e.clientX,
        y: e.clientY,
    };
}

export function findParentByClassName<T>(element, className): T | null {
    try {
        while (element && element) {
            const classes = element.getAttribute('class');
            if (classes && classes.indexOf(className) > -1) {
                return element;
            }
            element = element.parentElement;
        }
    } catch (e) {
        throw new Error(`findParentByClassName error: ${e}`);
    }
    return null;
}

export function triggerEvent(trigger: TriggerEvent) {
    switch (trigger) {
        case 'click': {
            return 'onClick';
        }
        case 'hover': {
            return 'onMouseOver';
        }
        case 'contextmenu': {
            return 'onContextMenu';
        }
        default: {
            return trigger;
        }
    }
}

/**
 * Get the element position by placement and DOMRect
 * @param placement top | right | bottom | left
 * @param domRect Dom rect info, normally get it from getBoundingClientRect function
 */
export function getPositionByPlacement(
    placement: PlacementType,
    domRect: DOMRect
): IPosition {
    let x = domRect.x; // Initial placement is top
    let y = domRect.y;

    if (placement === 'top') {
        y = domRect.y - domRect.height;
    } else if (placement === 'right') {
        x = domRect.x + domRect.width;
    } else if (placement === 'bottom') {
        y = domRect.y + domRect.height;
    } else if (placement === 'left') {
        x = domRect.x - domRect.width;
    } else if (placement === 'rightBottom') {
        x = domRect.x + domRect.width;
        y = domRect.y + domRect.height;
    }
    return { x, y };
}

export function getAttr(domElement: HTMLElement, attr) {
    return domElement.getAttribute(attr) || '';
}

/**
 * Get an element the center coords
 * @param element HTMLElement
 * @returns
 */
export function getElementClientCenter(element: HTMLElement) {
    const { left, top, width, height } = element.getBoundingClientRect();
    return {
        x: left + width / 2,
        y: top + height / 2,
    };
}

/**
 * Get the data-* attributions from props
 * @param props
 * @returns
 */
export function getDataAttributionsFromProps(
    props: Record<string, any>
): Record<string, any> {
    return Object.keys(props).reduce((pre, cur) => {
        if (cur.startsWith('data-')) {
            pre[cur] = props[cur];
        }
        return pre;
    }, {});
}
