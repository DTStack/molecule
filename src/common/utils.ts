export function searchById(id) {
    return (item) => item.id === id;
}

/**
 * Clone a new object by an object instance
 * @param origin Original object instance
 */
export function cloneInstance<T>(origin: T) {
    try {
        const prototypes = Object.getPrototypeOf(origin) || {};
        Object.keys(prototypes).forEach((prop) => {
            if (typeof prototypes[prop] === 'function') {
                prototypes[prop].bind(origin);
            }
        });
        return {
            ...origin,
            ...prototypes,
        };
    } catch (e) {
        console.error('Function cloneInstance error:', e);
    }
}

/**
 * Merge multiple functions to one function
 * @param funcs
 */
export function mergeFunctions(...funcs) {
    return function (...args) {
        funcs.filter((fn) => !!fn).forEach((fn) => fn?.(...args));
    };
}

export function randomId() {
    return Date.now() + Math.round(Math.random() * 1000);
}

export function mergeObjects<T>(source: object, target: object): T {
    for (const prop in target) {
        if (typeof source[prop] === 'object' && target[prop]) {
            source[prop] = mergeObjects(source[prop], target[prop]);
        } else {
            source[prop] = target[prop];
        }
    }
    return source as any;
}

/**
 * It's converts an object to a flatted object,
 * eg: { a: { b: 'test' }}, result is : { 'a.b': 'test' }
 * @param target flat target
 */
export function flatObject(target: object): object {
    const flatted = {};
    const recurse = (parent: string, target: object) => {
        for (const prop in target) {
            if (prop) {
                const value = target[prop];
                const path = parent ? `${parent}.${prop}` : prop;
                if (typeof value === 'object') {
                    recurse(path, value);
                } else {
                    flatted[path] = value;
                }
            }
        }
    };
    recurse('', target);
    return flatted;
}

/**
 * It's converts a flatted object to a normal object,
 *  eg: { 'a.b': 'test' }, result is : { a: { b: 'test' }}
 * @param target flat target
 */
export function normalizeFlattedObject(target: object): object {
    let normalized = {};

    for (const prop in target) {
        if (prop) {
            const flattedProps = prop.split('.');
            let prevProp = '';
            let root = {};
            let prevObj = root;
            for (let i = 0; i < flattedProps.length; ++i) {
                if (i === flattedProps.length) break;
                const key = flattedProps[i];
                const current = { [key]: '' };

                if (prevProp) {
                    prevObj[prevProp] = current;
                }
                prevObj = current;
                prevProp = key;
                if (i === 0) {
                    root = current;
                }
            }
            prevObj[prevProp] = target[prop];
            normalized = mergeObjects(normalized, root);
        }
    }
    return normalized;
}

/**
 * Determine if a color is light or dark.
 * @param color HEX or RGB
 */
export function colorLightOrDark(color: string) {
    // Variables for red, green, blue values
    let r: number;
    let g: number;
    let b: number;

    // Check the format of the color, HEX or RGB?
    if (color.match(/^rgb/)) {
        // If RGB --> store the red, green, blue values in separate variables
        const matchArray =
            color.match(
                /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
            ) || [];
        r = +matchArray[1];
        g = +matchArray[2];
        b = +matchArray[3];
    } else {
        // If hex --> Convert it to RGB
        let rgbNum = +('0x' + color.slice(1, 7));
        if (color.length < 5) {
            rgbNum = +('0x' + color.slice(1).replace(/./g, '$&$&').slice(0, 6));
        }
        r = rgbNum >> 16;
        g = (rgbNum >> 8) & 255;
        b = rgbNum & 255;
    }

    // HSP (Highly Sensitive Poo) equation from http://alienryderflex.com/hsp.html
    const hsp = Math.sqrt(0.299 * (r * r) + 0.587 * (g * g) + 0.114 * (b * b));

    // Using the HSP value, determine whether the color is light or dark
    if (hsp > 127.5) {
        return 'light';
    }
    return 'dark';
}
