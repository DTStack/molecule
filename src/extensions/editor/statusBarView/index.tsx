import * as React from 'react';

export function EditorStatusBarView(props: any) {
    const { data = { ln: 0, col: 0 } } = props;
    return <span>{`Ln ${data.ln}, Col ${data.col}`}</span>;
}
