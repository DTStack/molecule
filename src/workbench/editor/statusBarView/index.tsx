import React, { memo } from 'react';
import { IStatusBarItem } from 'mo/model/workbench/statusBar';
import { localize } from 'mo/i18n/localize';

export function EditorStatusBarView(props: IStatusBarItem) {
    const { data = { ln: 0, col: 0 } } = props;
    return (
        <span>
            {localize(
                'editor.singleSelection',
                'Line ${i}, Column ${i}',
                data.ln,
                data.col
            )}
        </span>
    );
}
export default memo(EditorStatusBarView);
