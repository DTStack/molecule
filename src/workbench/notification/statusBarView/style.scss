@import 'mo/style/common';

#{$bell} {
    height: 100%;
    position: relative;

    &::before {
        vertical-align: sub;
    }

    &--active {
        &::after {
            border: 5px solid;
            border-color: transparent transparent var(--statusBar-background);
            content: '';
            height: 0;
            left: 2px;
            position: absolute;
            top: -10px;
            width: 0;
        }
    }
}
