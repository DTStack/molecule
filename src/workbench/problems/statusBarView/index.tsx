import React from 'react';
import { Icon } from 'mo/components/icon';
import { IStatusBarItem } from 'mo/model/workbench/statusBar';

export function ProblemsStatusBarView(props: IStatusBarItem) {
    const { data = { errors: 0, warnings: 0, infos: 0 } } = props;
    return (
        <>
            <Icon type="error" />
            {` ${data.errors} `}
            <Icon type="warning" />
            {` ${data.warnings} `}
            <Icon type="info" />
            {` ${data.infos}`}
        </>
    );
}
export default React.memo(ProblemsStatusBarView);
