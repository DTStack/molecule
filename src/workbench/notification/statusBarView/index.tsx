import React, { memo, useEffect, useRef } from 'react';
import { Icon } from 'mo/components/icon';
import type { INotification } from 'mo/model';
import { NotificationPane } from '../notificationPane';
import type { INotificationController } from 'mo/controller';
import { classNames, getBEMModifier, prefixClaName } from 'mo/common/className';
import { select } from 'mo/common/dom';
import { render } from 'mo/react/render';

const defaultBellClassName = prefixClaName('bell');
const activeBellClassName = getBEMModifier(defaultBellClassName, 'active');

export function NotificationStatusBarView(
    props: INotification & Partial<INotificationController>
) {
    const {
        data = [],
        onClick,
        showNotifications,
        id,
        actionBar,
        onActionBarClick,
        onCloseNotification,
    } = props;
    const wrapper = useRef<HTMLDivElement>();
    const hasNotifications = data.length > 0;
    const renderIcon = hasNotifications ? 'bell-dot' : 'bell';

    useEffect(() => {
        const container = select(`.${prefixClaName('workbench')}`);
        if (container) {
            wrapper.current = wrapper.current || document.createElement('div');
            container.appendChild(wrapper.current);
            render(
                <NotificationPane
                    id={id}
                    data={data}
                    actionBar={actionBar}
                    showNotifications={showNotifications}
                    onActionBarClick={onActionBarClick}
                    onCloseNotification={onCloseNotification}
                />,
                wrapper.current
            );
        }
    }, [
        id,
        data,
        actionBar,
        showNotifications,
        onActionBarClick,
        onCloseNotification,
    ]);

    return (
        <Icon
            className={classNames(
                defaultBellClassName,
                showNotifications && activeBellClassName
            )}
            onClick={onClick}
            type={renderIcon}
        />
    );
}
export default memo(NotificationStatusBarView);
