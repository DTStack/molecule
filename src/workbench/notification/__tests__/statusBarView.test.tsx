import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { NotificationStatusBarView } from '../statusBarView';
import { select } from 'mo/common/dom';
import { expectFnCalled } from '@test/utils';

describe('Test Notification StatusBar View Component', () => {
    test('Match The NotificationStatusBarView snapshot', () => {
        const component = renderer.create(
            <NotificationStatusBarView
                data={[{ id: '', value: '' }]}
                id="test"
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Should display the bell dot icon', () => {
        const { rerender } = render(<NotificationStatusBarView id="test" />);
        expect(select('.codicon-bell-dot')).not.toBeInTheDocument();
        rerender(
            <NotificationStatusBarView
                id="test"
                data={[{ id: '', value: '' }]}
            />
        );
        expect(select('.codicon-bell-dot')).toBeInTheDocument();
    });

    test('Listen to the onClick event', () => {
        expectFnCalled((fn) => {
            const { container } = render(
                <NotificationStatusBarView id="test" onClick={fn} />
            );
            fireEvent.click(container.firstElementChild!);
        });
    });
});
