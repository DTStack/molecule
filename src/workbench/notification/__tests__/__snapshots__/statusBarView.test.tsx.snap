// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Notification StatusBar View Component Match The NotificationStatusBarView snapshot 1`] = `
Array [
  <span
    className="codicon codicon-bell-dot"
  />,
  <div
    className="mo-notification mo-context-view--shadow"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <header
      className="mo-notification__header"
    >
      <span>
        Notifications
      </span>
      <div
        className="mo-action-bar"
      >
        <ul
          className="mo-action-bar__container"
        />
      </div>
    </header>
    <div
      className="mo-notification__body"
    >
      <div>
        
        <span
          className="mo-notification--close codicon codicon-close"
          onClick={[Function]}
          title="Clear Notification"
        />
      </div>
    </div>
  </div>,
]
`;
